### ISSUE TITLE
Spotweb 504 error when using the API as indexer

### ISSUE THREAD

**Issue by tharealmb** [spotweb/spotweb#346]
I set up my Spotweb installation on a Synology NAS. Spotweb is working fine when accessing it directly.

I also added Spotweb as an Indexer in Sonarr and Couchpotato. But it won't use Spotweb as an indexer because the programs get a time-out when trying to acces Spotweb using the API.

When i test the connection in Sonarr, all seems well. I can test the connection and it says its working. 
But when i try to do a search i get an error:

Newznab   Spotweb HTTP request failed: [504:GatewayTimeout] [GET] at [http://localhost/spotweb/api?t=tvsearch&cat=5000&extended=1&apikey=(removed)&offset=0&limit=50&rid=23792&tvmazeid=1228&season=5&ep=5]   16:10
HttpClient   HTTP Error - Res: [GET] http://localhost/spotweb/api?t=tvsearch&cat=5000&extended=1&apikey=(removed)&offset=0&limit=50&rid=23792&tvmazeid=1228&season=5&ep=5: 504.GatewayTimeout   16:10
NzbSearchService   Searching 2 indexers for [MyTVSeries : S05E05]  16:09

If i visit the above link manually and add my APIkey from Spotweb, i also get a 504 error page.
If i visit localhost/spotweb i do get Spotweb, can update and download files. So it seems API specific to me (as noob).

![bsuqzof](https://user-images.githubusercontent.com/6922507/33651794-5b526436-da67-11e7-89ef-44601c861679.png)

**Comment by mesa57** [comment]
The query which is executed takes more then 16 secs on my test Synology dsm vm.

**Comment by Sweepr** [comment]
The 504 Gateway Timeout error is an HTTP status code that means that one server did not receive a timely response from another server that it was accessing while attempting to load the web page or fill another request by the browser.

**Comment by Sweepr** [comment]
On 2 different machine's the query run's almost instantly, less then 1,5 second from request to response.

Seems to be Synology related once again?

**Comment by Sweepr** [comment]
Do all requests take this long?

**Comment by tharealmb** [comment]
I do get the error on all requests. It did work a few times but that was more exception. Since then it hasn't worked once.

**Comment by mesa57** [comment]
@Sweepr : you have a Synlogy, can you test there ?
I only have a Synlogy VM, which shows the same problem.

**Comment by tharealmb** [comment]
Just did a test using couchpotato and Sonarr. If you have something you want me to test specifically let me know.
Couchpotato:
Log gives the following message:
```
12-07 09:42:21 INFO
[hpotato.core.plugins.base] Opening url: get https://api.couchpota.to/eta/tt2310332/, data: []

12-07 09:42:22 INFO
[core.media.movie.searcher] Search for The Hobbit: The Battle of the Five Armies in 2160p ignoring ETA

12-07 09:42:22 INFO
[hpotato.core.plugins.base] Opening url: get http://192.168.1.200/spotweb/api?imdbid=2310332&apikey=xxx&t=movie&extended=1&cat=2000, data: []
```
When i try to open the requested URL from Spotweb http://192.168.1.200/spotweb/api?imdbid=2310332&apikey=xxx&t=movie&extended=1&cat=2000 i also get a timeout after exactly one minute (probably a setting in Chrome)

For sonarr:
```
Searching 1 indexers for [Game of Thrones : S07E01]

HTTP Error - Res: [GET] http://localhost/spotweb/api?t=tvsearch&cat=5000&extended=1&apikey=(removed)&offset=0&limit=250&rid=24493&tvmazeid=82&season=7&ep=1: 504.GatewayTimeout
Close

Spotweb HTTP request failed: [504:GatewayTimeout] [GET] at [http://localhost/spotweb/api?t=tvsearch&cat=5000&extended=1&apikey=(removed)&offset=0&limit=250&rid=24493&tvmazeid=82&season=7&ep=1]
Close

```
Here the same. If i directly open the link in Chrome, i get the same result. 504 error after 1 minute.

If you want me to do more specific tests let me know.

**Comment by MrLurch81** [comment]
I used to experience this issue as well (Synology DS1512+). For me the timeouts were caused by the resulting DB query being 'very heavy' for MariaDB (as shown by a high disk-IO during the API call).

Moving the MariaDB datafiles to a SSD helped as a workaround, but still... There is room to improve these queries.

Testing your URL (http://<host>/spotweb/api?imdbid=2310332&apikey=xxx&t=movie&extended=1&cat=2000) in my current configuration takes ~8 sec.

**Comment by mesa57** [comment]
Because searching via imdbid search the title in English as wel as in Dutch this leads to a very expensive OR query. MySql FTS is not capable of handling OR query's. On my syno VM this query took 180 secs.
This cannot be solved.

**Comment by tharealmb** [comment]
Ok. That means for me I'll give up on using spotweb as an indexer, since I can't force couchpotato and sonarr to do name queries. I was thinking of moving everything to a raspberry pi but if it's a disk usage problem it'll be the same or worse. Thanks for looking into this though, at least I know what the issue is.

**Comment by mesa57** [comment]
Fixes long response when requesting for a specific season and ep.
For imdb query's the noalt=1 parameter is added.
That prevents searching for alternative titles, therefore prevents an OR query and accelarates response.
noalt=1 also works when a season search is done. In that case no search on 'Seizoen x' is done and also prevents an OR query.

**Comment by MrLurch81** [comment]
Just did some more research into MySQL FTS.
Are you required to perform both "s.title LIKE" as well as "MATCH (s.title) AGAINST ..." in your where clause?
When the FTS matching is combined into one MATCH (it supports OR matches using parentheses):
`MATCH (s.title) AGAINST (' +("Moviename Dutch" "Moviename English") +(2014) ) ' IN BOOLEAN MODE)`
this will require presence of either the Dutch or English name in the file, as well as the year.
After removing the "s.title LIKE" statements, the explain plan improves a lot, which is also reflected in the query performance (reduced from ~13 to 2 sec on my Syno+SSD drive).

There is an associated change in the result set (from 10 to 31 rows). This is due to dropping the LIKE statements, which also allows results without the ":" (colon) in this specific movie name to be included in the result set. All 31 rows seem to be correct results.

So, to summarize the query changes I looked at: Removed all LIKE matching on s.title and merged both languages into one single MATCH statement. Of course this also means updating the searchrelevancy0 and removing the searchrelevancy1 column from the SELECT and ORDER BY.

This is the resulting WHERE clause:
`WHERE bl.spotterid IS NULL
	AND s.category = 0
	AND s.subcatz = 'z0|'
	AND s.filesize > 0
	AND MATCH (s.title) AGAINST ('+("Moviename Dutch" "Moviename English") +(2014)' IN BOOLEAN MODE)`

**Comment by mesa57** [comment]
The code which generates the match / likes does not support those constructs. However, I made some enhancements, see develop branch.

**Comment by MrLurch81** [comment]
Thanks for your reply, @mesa57. Yes indeed I noticed the changes in your develop branch. However, as far as I understood them, they require a client side change as well? Of course it would probably be the cleanest solution if the client does not have to be aware of including/excluding alternative titles?

So, in the meantime, I have been trying to modify the `dbfts_mysql.php` file to merge multiple MATCH constructs for the same table into one `filterValueSql`. Although this does work, it leaves me with a question about the original way the WHERE statement is constructed (for MySQL): in some situations, the searchMode is set to 'both-...', so there is both a LIKE and a MATCH statement for the same table in the WHERE clause.
Could you perhaps elaborate why in some cases a "both-" search is done and why in this situation an AND is used (instead of an OR) to combine the LIKE statements with the MATCH statement? It would seem logical to switch to a LIKE statement if a MATCH statement would not produce any results, but I would guess that if a result is already found using LIKE, the resultset will likely not be limited any further by combining it with a MATCH using AND?

As an example for imdbid=2310332 ("The Hobbit: The Battle of the Five Armies"), the English Title searchItem has both Stop and Non-Stop words, as well as both TooShort and LongEnough words. So I would expect it to not have to fall-back to LIKE but just use MATCH.
Under the original mechanism ("both-match-boolean") using both languages, 10 results are given, in around 13 secs. When temporarily disabling the fall-back to "both-..." and merging the MATCH statements, the query is performed using "match-boolean" and the results are both more complete (the same 10 results, plus an additional 21 accurate results that do not have the colon present in the movie title) and much faster: <0,1 sec.

**Comment by mesa57** [comment]
It is the way around. The match retrieves all rows containing on of the words, The LIKE is used to limit the search results. Btw. : I do not fully understand how dbfts-mysql works. But I know that changes should also work for postgresql and sqlite.

**Comment by MrLurch81** [comment]
Thanks for the update. Indeed dbfts-mysql is a complicated matter, the previous adjustments I made (in fork) worked fine for the imdb query, but gave too many results for a series query (e.g. from Sonarr).
Perhaps I'll be able to work on those situations later.

**Comment by mesa57** [comment]
Basicly a good mechanism to process OR / AND in match query's should be build. 
I think use of parenthesis is necessary. 
The condition stack spotweb uses is in my opinion a poor translation (but needed to support multiple databases).
